### PR TITLE
Implement partial retraining

### PR DESCRIPTION
**Pull request by ashelkovnykov**
These changes introduce an option to specify a path to an existing GAME model and a set of "locked" coordinates from that model. These locked coordinates are re-used from the pre-trained model: the training and validation data scores using these coordinates from the pre-trained model are used during training of the new model.

This allows users to retrain a component of an existing model, or to use model components elsewhere in other models.

Unit and integration tests added. Tests on live data show that partially retrained models produce coefficients and evaluation metrics within 1E-6 of the original model, when run on the same data.

### CODE REVIEW

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/GameTrainingDriverIntegTest.scala [hunk 1]
```diff
@@ -347,6 +347,32 @@ class GameTrainingDriverIntegTest extends SparkTestUtils with GameTestUtils with
...
+    runDriver(partialRetrainWithFixedBaseArgs.put(GameTrainingDriver.rootOutputDirectory, outputDir))
+
+    compareModelEvaluation(new Path(outputDir, "best"), trainedMixedModelPath, TOLERANCE)
+  }
+
+  /**
+   * Test GAME partial retraining using a pre-trained random effects model.
+   */
+  @Test
+  def testPartialRetrainWithRandomBase(): Unit = sparkTest("testPartialRetrainWithFixedBase", useKryo = true) {
```
Comment (alice2008) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/GameTrainingDriverIntegTest.scala]: naming is wrong: should be `testPartialRetrainWithRandomBase`

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/GameTrainingDriverIntegTest.scala [hunk 2]
```diff
@@ -347,6 +347,32 @@ class GameTrainingDriverIntegTest extends SparkTestUtils with GameTestUtils with
...
+
+    val outputDir = new Path(getTmpDir, "testPartialRetrainWithRandomBase")
+
+    runDriver(partialRetrainWithFixedBaseArgs.put(GameTrainingDriver.rootOutputDirectory, outputDir))
```
Comment (alice2008) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/GameTrainingDriverIntegTest.scala]: should be `partialRetrainWithRandomBaseArgs`

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala [hunk 3]
```diff
@@ -248,6 +308,13 @@ class GameEstimator(val sc: SparkContext, implicit val logger: Logger) extends P
     // Verify valid GameEstimator settings
     validateParams()
 
+    // Locked coordinates should not be present in the optimization configurations
```
[the remark this answers is not in the document]
Comment (ashelkovnykov) [reply photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: I think it makes sense that the locked coordinates shouldn't have coordinate configurations explicitly specified for the following reasons:
- We don't need this information - we're enforcing that user configuration files should not have unnecessary and potentially confusing information
- If we need it, we can get it from the metadata stored with each model
- For the above reason, we would want to avoid mismatches between historic and new configuration data

I'm okay with modifying this if need be later, though for now I think we should do it this way. I think it's better to be too strict and become more lenient, than to be too lenient from the start.

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala [hunk 4]
```diff
@@ -580,58 +653,66 @@ class GameEstimator(val sc: SparkContext, implicit val logger: Logger) extends P
...
-          val treeAggDepth = if (feDataSet.numFeatures < FIXED_EFFECT_FEATURE_THRESHOLD) {
-            getOrDefault(treeAggregateDepth)
-          } else {
-            Math.max(getOrDefault(treeAggregateDepth), DEEP_TREE_AGGREGATE_DEPTH)
+    val coordinates = getRequiredParam(coordinateUpdateSequence)
+      .map { coordinateId =>
+
+        val lockedCoordinates = get(partialRetrainLockedCoordinates).getOrElse(Set())
+        val coordinate: Coordinate[_] = if (lockedCoordinates.contains(coordinateId)) {
+          trainingDataSets(coordinateId) match {
```
[the remark this answers is not in the document]
Comment (ashelkovnykov) [reply photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: That's right - this is so that we can compute the residuals and add them to the offsets of the coordinates that we're training.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala [hunk 5]
```diff
@@ -126,17 +216,18 @@ class CoordinateDescent(
     fullTrainingScore.persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL).materialize()
 
     // Initialize the regularization term value
-    var regularizationTermValueContainer = coordinates
-      .map { case (coordinateId, coordinate) =>
-        (coordinateId, coordinate.computeRegularizationTermValue(updatedGameModel.getModel(coordinateId).get))
+    var regularizationTermValueContainer = coordinatesToTrain
```
[the remark this answers is not in the document]
Comment (ashelkovnykov) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala]: Yes, that's a piece of information that we lose. The regularization is already baked into the coefficients from the trained model, so we can't access it. Thinking about it, this results in misleading logs about objective function vs. regularization term loss values. I'll add a 'TODO' about this, but won't resolve it for now.

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: Thanks for the explanation. Should be fine to leave it for now.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (alice2008) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala]: I am thinking it wouldn't be hard to compute the regularization term from the trained model. I am worried about the consistency of different training. For regular case, all coordinates would have regularization term contribute to the overall loss. Here, only `trainedCoordinates` would be part of the regularization term. In partial training case, will we get into more **overfitting** situation?

[the remark this answers is not in the document]
Comment (ashelkovnykov) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala]: This `fullRegularizationTerm` is used only for logging purposes, it's not a part of the optimization algorithm. The regularization term used during optimization is a trait that's applied to the `ObjectiveFunction` class, which modifies the behaviour of the `value`, `gradient`, and `hessian` functions. Take a look at where `regularizationTermValueContainer` and the `updatedScoresContainer` are used in `CoordinateDescent` and you'll see what I mean.

The consistency between training should remain the same, as the regularization terms do not contribute to the residuals applied to the offsets of whichever coordinate is being trained (see line 189 of `CoordinateDescent`).

**Comment by li-ashelkov** [comment]
Ran `git rebase`, tests shouldn't have been affected. If they were, I have more PRs coming which will address any failures this merge may create.
